import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/CStoreService.java

```diff
@@ -72,12 +81,42 @@ protected void store(
           DicomStreamUtil.dicomStreamWithFileMetaHeader(
               sopInstanceUID, sopClassUID, transferSyntax, countingStream);
 
-      dicomWebClient.stowRs(path, inBuffer);
+      if (redactor != null) {
+        PipedOutputStream pipedOut = new PipedOutputStream();
+        PipedInputStream pipedIn = new PipedInputStream(pipedOut);
 
-      log.info("Received C-STORE for association {}, SOP class {}, TS {}, remote AE {}",
-          association.toString(), sopClassUID, transferSyntax, remoteAeTitle);
-      response.setInt(Tag.Status, VR.US, Status.Success);
+        CompletableFuture<Throwable> futureThrowable = new CompletableFuture<>();
+        association.getApplicationEntity().getDevice().execute(() -> {
+          try {
+            dicomWebClient.stowRs(path, pipedIn);
+            futureThrowable.complete(null);
+          } catch (Throwable e) {
+            futureThrowable.complete(e);
+            try {
+              pipedIn.close();
+            } catch (IOException e1) {
+              log.error("Failed to close pipedIn", e);
+            }
```

Comment:
Should this be in a finally {} block? Do we always want to close pipedIn?

Reply:
redact has both input and output in try-with-resources, so ultimately closing streams is not in question. Closing it from here is done to intentionally fail redact (as opposed to having it wait indefinitely after stowRs already failed).

Reply:
I see, can you please add a small comment explaining this?

Reply:
Ok